lib/block_view_atomic_txns_test.go

```diff
@@ -0,0 +1,388 @@
+
+	//
+	// Now we test that the transactions are truly dependent on each-other by reorganizing them.
+	//
+
+	// Reorganize the transactions.
+	initialTxn := atomicTxns[0]
+	atomicTxns[0] = atomicTxns[len(atomicTxns)-1]
+	atomicTxns[0] = initialTxn
+
+	// Initialize test chain, miner, and testMeta for failing use.
+	testMetaFail := _setUpMinerAndTestMetaForAtomicTransactionTests(t)
+
+	// Initialize m0, m1, m2, m3, m4.
+	_setUpUsersForAtomicTransactionsTesting(testMetaFail)
+
+	// Construct a new view to connect the transactions to.
+	utxoView, err = NewUtxoView(
+		testMetaFail.db, testMetaFail.params, testMetaFail.chain.postgres, testMetaFail.chain.snapshot, nil)
+	require.NoError(t, err)
+	blockHeight = testMetaFail.chain.BlockTip().Height + 1
+
+	// Connect the transactions to ensure they can actually be connected.
+	for _, txn := range atomicTxns {
+		// Connect the transaction.
+		txHash := txn.Hash()
+		_, _, _, _, err := utxoView.ConnectTransaction(
+			txn, txHash, blockHeight, 0, true, false)
+		if err != nil {
+			require.Contains(t, err.Error(), RuleErrorInsufficientBalance)
+		}
+	}
+}
+
+//----------------------------------------------------------
+// (Testing) Atomic Transactions Setup Helper Functions
+//----------------------------------------------------------
+
+// The goal of _generateDependentAtomicTransactions is to generate
+// a sequence of transactions who CANNOT be reordered meaning they
+// must be executed in the sequence returned. This mean transaction
+// with position ii in atomicTransactions CANNOT be placed in an
+// index jj of atomicTransactions such that jj < ii
+//
+// How can we generate "dependent atomic transactions" algorithmically using
+// TestMeta initialized with _setUpUsersForAtomicTransactionTesting?
+//
+//	(1) Choose an arbitrary starter public key with DESO (m0PkBytes)
+//	(2) For ii in [0, numberOfTransactions):
+//		(2a) Generate a new public/private key pair (pub_ii, priv_ii)
+//		(2b) Have pub_(ii-1) do a max DESO transfer to pub_ii. Use m0PkBytes as pub_(-1).
+//	(3) Have pub_numberOfTransactions perform a max transfer back to m0PkBytes
+//
+// Notice that because pub_ii only has DESO at following the transaction with
+// the iith index in numberOfTransactions, it's impossible to reorder the transactions
+// in any other order. Hence, these transactions are dependent on each other.
+//
+// The length of the returned list of transactions is specified by numberOfTransactions.
+func _generateDependentAtomicTransactions(
+	testMeta *TestMeta,
+	numberOfTransactions int,
+) (
+	_atomicTransactions []*MsgDeSoTxn,
+) {
+	var atomicTransactions []*MsgDeSoTxn
+	var receiverPublicKeysBase58 []string
+	var receiverPrivateKeysBase58 []string
+	var receiverBalancesNanos []uint64
+
+	// Get the initial balance of m0.
+	m0InitialBalanceNanos := _getBalance(
+		testMeta.t, testMeta.chain, testMeta.mempool, m0Pub)
+
+	// Generate the atomic transactions.
+	for ii := 0; ii < numberOfTransactions-1; ii++ {
```

Comment:
Why `numberOfTransactions - 1`? This means we'll have `numberOfTransactions - 1` in the list of innerTxns, which seems incorrect.

Reply:
ah okay it's the max transfer at the end.